core/kernel/interrupt.c

```diff
@@ -35,15 +35,23 @@ static struct itr_handler *find_handler(size_t it)
 
 void itr_handle(size_t it)
 {
-	struct itr_handler *h = find_handler(it);
+	struct itr_handler *h = NULL;
+	enum itr_return ret = ITRR_NONE;
 
-	if (!h) {
-		EMSG("Disabling unhandled interrupt %zu", it);
-		itr_chip->ops->disable(itr_chip, it);
-		return;
+	SLIST_FOREACH(h, &handlers, link) {
+		if (h->it != it)
+			continue;
+
+		ret = h->handler(h);
+
+		if (ret != ITRR_HANDLED && !(h->flags & ITRF_SHARED)) {
+			EMSG("Disabling interrupt %zu not handled by handler", it);
+			itr_chip->ops->disable(itr_chip, it);
+			return;
+		}
 	}
 
-	if (h->handler(h) != ITRR_HANDLED) {
+	if (ret != ITRR_HANDLED) {
```

Comment:
If it's handled by at least one handler we should keep the interrupt enabled.

Reply:
> 
> 
> If it's handled by at least one handler we should keep the interrupt enabled.

this is out for SLIST_FOREACH() , means no any handler can really handle this irq , just as original interrupt.c , would make the interrupt disable.

Reply:
An example:
There's two handlers registered as `ITRF_SHARED` for the same interrupt number.
The first handler sees that its device has something pending and deals with it and returns `ITRR_HANDLED`, this is stored in `ret`.
The second handler doesn't see anything pending on its device and returns `ITRR_NONE`, now `ret` is overwritten.
This leads to the interrupt becoming disabled here.

Reply:
got it , 
how about change as below , use or operator to keep the history ? 

```
 void itr_handle(size_t it)
 {
    struct itr_handler *h = NULL;
@@ -42,7 +32,7 @@ void itr_handle(size_t it)
        if (h->it != it)
            continue;

-       ret = h->handler(h);
+       ret |= h->handler(h);

        if (ret != ITRR_HANDLED && !(h->flags & ITRF_SHARED)) {
            EMSG("Disabling interrupt %zu not handled by handler", it);
```

Reply:
I'd prefer a helper variable `bool was_handled = false` to be updated instead relying on the bit pattern behind these enums.

Reply:
OK , v3 is pushed , please review. 

Thanks,
Davidwang